Core/include/Acts/EventData/GenericParticleHypothesis.hpp

```diff
@@ -0,0 +1,137 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/PdgParticle.hpp"
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/EventData/Charge.hpp"
+
+#include <utility>
+
+namespace Acts {
+
+/// @brief Particle hypothesis used in reconstruction
+///
+/// Our reconstruction hypothesis consists of absolute PDG code, mass and
+/// absolute charge
+template <typename charge_t>
+class GenericParticleHypothesis {
```

Comment:
Will try to use private inheritance from a detail struct and `using` to specialize and include only the factories that apply to a charge class.